counterpartylib/lib/messages/versions/enhanced_send.py

```diff
@@ -0,0 +1,202 @@
+#! /usr/bin/python3
+
+import struct
+import json
+import logging
+logger = logging.getLogger(__name__)
+
+from counterpartylib.lib import (config, util, exceptions, util, message_type, address)
+
+FORMAT = '>QQ21s'
+LENGTH = 8 + 8 + 21
+MAX_MEMO_LENGTH = 34
+ID = 2 # 0x02
+
+def unpack(db, message, block_index):
+    try:
+        # account for memo bytes
+        memo_bytes_length = len(message) - LENGTH
+        if memo_bytes_length < 0:
+            raise exceptions.UnpackError('invalid message length')
+        if memo_bytes_length > MAX_MEMO_LENGTH:
+            raise exceptions.UnpackError('memo too long')
+
+        struct_format = FORMAT + ('{}s'.format(memo_bytes_length))
+        asset_id, quantity, short_address_bytes, memo_bytes = struct.unpack(struct_format, message)
+        if len(memo_bytes) == 0:
+            memo_bytes = None
+        
+        # unpack address
+        try:
+          full_address = address.unpack(short_address_bytes)
+        except:
+          raise exceptions.UnpackError('address invalid')
+
+        # asset id to name
+        asset = util.generate_asset_name(asset_id, block_index)
+        if asset == config.BTC:
+          raise exceptions.AssetNameError('{} not allowed'.format(config.BTC))
+
+    except (struct.error) as e:
+        logger.warning("enhanced send unpack error: {}".format(e))
+        raise exceptions.UnpackError('could not unpack')
+
+    except (exceptions.AssetNameError, exceptions.AssetIDError) as e:
+        logger.warning("enhanced send invalid asset id: {}".format(e))
+        raise exceptions.UnpackError('asset id invalid')
+
+    unpacked = {
+      'asset': asset,
+      'quantity': quantity,
+      'address': full_address,
+      'memo': memo_bytes,
+    }
+    return unpacked
+
+def validate (db, source, destination, asset, quantity, memo_bytes, block_index):
+    problems = []
+
+    if asset == config.BTC: problems.append('cannot send {}'.format(config.BTC))
+
+    if not isinstance(quantity, int):
+        problems.append('quantity must be in satoshis')
+        return problems
+
+    if quantity < 0:
+        problems.append('negative quantity')
+
+    # if quantity == 0:
+    #     problems.append('zero quantity')
```

Comment:
I guess the old `send` allowed to send 0 of an asset... weird ...  
I think it makes sense to make it require > 0 so uncomment this (and add a test to match the failure case)